File: audiocomposer.js
// variables for testing 
const bucket = "storycorps-signature-remote"
const account = "46565552"
const interview = "f105311e-e144-49fb-87ae-b63a341a18fe"
const interviewId = "sdhTest 2020-4-16 23:59:17"
const name = interview

// Load the SDK and UUID
var AWS = require('aws-sdk');
// Create an S3 client
var s3 = new AWS.S3({apiVersion: '2006-03-01'});
var params = {Bucket: bucket, Key: zippedKey};
tmp_file_name = './tmp/' + name + '/archive.zip'
fs.mkdir('./tmp/' + name, { recursive: true }, (err) => {
  if (err) throw err;
});
fs.writeFile(tmp_file_name, "Hello World!", function(err) {
  if(err) {
      return console.log(err);
  }
});

var temp_dir = './tmp/' + name;



var fs = require("fs"),
  exec = require("child_process").execSync,
  fs = require("fs"),
  program = require("commander"),
  path = require("path"),
  unzip = require("unzip-stream"),
  child;

const zippedKey = account + "/" + interview + "/archive.zip"
const unzippedLocation = account + "/" + interview

var errHandler = function(err) {
  console.log(err);
}

async function downloadFile (params, tmp_file_name){
  return new Promise((resolve, reject) => {
    console.log("downloading file");
    var file = require('fs').createWriteStream(tmp_file_name);
    resolve(await s3.getObject(params).createReadStream().pipe(file));
    // resolve(result);
  })
}

async function unzipFile(tmp_file_name, unzip, temp_dir){
  return new Promise((resolve, reject) => {
    console.log("unzipping");
    var unzipped = await fs.createReadStream(tmp_file_name).pipe(unzip.Extract({ path: temp_dir }))
    resolve(unzipped);
  })
}

function convert(tmp_file_name, unzip, temp_dir){
  return new Promise((resolve,reject) => {
      var files = fs.readdirSync(temp_dir);

      var json_file;
      files.forEach(function (file) {
        if (path.extname(file) == ".json") {
          json_file = file;
        }
      });

      if (!json_file) {
        console.log("ZIP file does not contain a json file");
      }

      var script = JSON.parse(
        fs.readFileSync(path.join(temp_dir, json_file)).toString()
      );

      var archiveId = script.id;
      var archive_path = temp_dir;
      var format = program.format;

      var startTime = 10000000000000;
      var endTime = 0;

      // find start end end time for the whole playback
      script.files.forEach(function (e) {
        if (e.startTimeOffset < startTime) {
          startTime = e.startTimeOffset;
        }
        if (e.stopTimeOffset > endTime) {
          endTime = e.stopTimeOffset;
        }
      });

      // make them all 0 based
      script.files.forEach(function (e) {
        e.startTimeOffset -= startTime;
        e.stopTimeOffset -= startTime;
      });

      // sort them by start time
      script.files.sort(function (a, b) {
        return a.startTimeOffset - b.startTimeOffset;
      });

      console.log("script=",script);
      console.log("duration=", endTime - startTime);

      var inputs = "";

      // Loop over the files to Create a WAV file for each path that has the same offset at the start.
      script.files.forEach((oneFile) => {
        
        let speaker_name = JSON.parse(oneFile.connectionData).userName.replace(/[^a-z0-9]/gi, '_');

        // generate a single wavefile with a delay at the front of it.
        let inputFile = `${archive_path}/${oneFile.filename}`;
        let outputFile = `${archive_path}/${speaker_name}-${oneFile.filename}.wav`;
        cmd = `ffmpeg -y -loglevel warning -acodec libopus -i ${inputFile} -af "adelay=${oneFile.startTimeOffset}|${oneFile.startTimeOffset}" ${outputFile}`;

        // console.log("individual command:\n", cmd);
        console.log("executing individual command for " + inputFile)

        child = exec(cmd, function (error, stdout, stderr) {
          if (error !== null) {
            console.log("exec error: " + error);
          }
        });
        uploadFile(archive_path + '/'+ oneFile.filename + '.wav', bucket, unzippedLocation + '/' + oneFile.filename + '.wav');

        // add to our list of inputs for the big merge
        inputs += ` -itsoffset ${oneFile.startTimeOffset} -acodec libopus -i ${inputFile} `;
        // inputs += ` -itsoffset ${oneFile.startTimeOffset} -i ${inputFile}.wav `;

      });

      // now mix it down into one wav file.
      cmd = `ffmpeg -y -loglevel warning ${inputs} -filter_complex amix=inputs=${script.files.length}:duration=longest:dropout_transition=3 ${archive_path}/mixed-${archiveId}.wav`;

      console.log("command:\n", cmd);

      child = exec(cmd, function (error, stdout, stderr) {
        if (error !== null) {
          console.log("Mixdown error: " + error);
        }
      });

      //todo: upload the mixed file

      // console.log(`\n\nDone Processing! Files can be found in ${archive_path}\n`)
      resolve(`\n\nDone Processing! Files can be found in ${archive_path}\n`)
  })
}

const uploadFile = async(filePath, bucketName, key) => {
  return new Promise((resolve, reject) => {
    console.log("starting file upload: " + filePath)
    fs.readFile(filePath, (err, data) => {
      if (err){
        reject(err)
      }
      var base64data = new Buffer(data, 'binary');
      var params = {
        Bucket: bucketName,
        Key: key,
        Body: base64data
      };
      s3.upload(params, (err, data) => {
        if (err) console.error(`Upload Error ${err}`);
        resolve('Upload Completed');
      });
    });
  })
}

async function main(){
  await downloadFile(params, tmp_file_name);
  await unzipFile(tmp_file_name, unzip, temp_dir);
  await convert;
}
main()

File: .vscode/audiocomposer.py
from __future__ import print_function

import urllib
import zipfile
import boto3
import io
import os
import json
import subprocess

#todo: parse out the interview ID from the name
#todo: set these variables based on the webhook 
bucket = "storycorps-signature-remote"
account = "46565552"
interview = "f105311e-e144-49fb-87ae-b63a341a18fe"
interviewId = "TLO333333"

s3_client = boto3.client('s3')
s3_resource = boto3.resource('s3')
my_bucket = s3_resource.Bucket(bucket)

zippedKey = account + "/" + interview + "/archive.zip"
unzippedLocation = account + "/" + interview
temp_dir = './tmp/' + interview
interviewId = interviewId.lower()

def lambda_handler(key):
    try:
        obj = s3_client.get_object(Bucket=bucket, Key=key)
        putObjects = []
        with io.BytesIO(obj["Body"].read()) as tf:
            # Read the file as a zipfile and process the members
            with zipfile.ZipFile(tf, mode='r') as zipf:
                for file in zipf.infolist():
                    fileName = account + "/" + interview + "/" + file.filename
                    putFile = s3_client.put_object(Bucket=bucket, Key=fileName, Body=zipf.read(file))
                    putObjects.append(putFile)

        # for each object in the bucket/account/archiveID directory
        objs = my_bucket.objects.filter(Prefix=unzippedLocation)
        #make a folder in tmp
        if not os.path.exists("tmp/" + interview):
                os.makedirs("tmp/" + interview)
        #download the files
        for obj in objs:
            if(obj._key[-5:] == ".json"):
                s3_client.download_file(bucket, obj.key, "tmp/" + interview + "/interview.json")
            elif(obj._key[-5:] == ".webm" == ".webm"):
                s3_client.download_file(bucket, obj.key, "tmp/" + obj.key[-78:])
            pass
        
        # open the JSON file 
        f = open('tmp/' + interview + '/interview.json',) 
        
        # returns JSON object as a dictionary 
        data = json.load(f)
        startTime = 10000000000000;
        endTime = 0;
        
        # find start end end time for the whole playback
        for file in data['files']: 
            if(file["startTimeOffset"] < startTime):
                startTime = file["startTimeOffset"]
            if(file["stopTimeOffset"] > endTime):
                endTime = file["stopTimeOffset"]

        for file in data['files']: 
            file["startTimeOffset"] -= startTime
            file["stopTimeOffset"] -= endTime
        # print("script=",data);
        # print("duration=", endTime - startTime);

        inputs = ""

        for file in data['files']: 
            speaker_name = json.loads(file["connectionData"])["userName"]
            # generate a single wavefile with a delay at the front of it.
            inputFile = temp_dir + "/" + file["filename"]
            fileName = speaker_name + "-" + file["filename"] + ".wav"
            outputFile = temp_dir + "/" + fileName

            cmd = "ffmpeg -y -loglevel warning -acodec libopus -i " + inputFile + " -af adelay=" + str(file["startTimeOffset"]) + " " + outputFile
            p = subprocess.call(cmd, shell=True)
            
            s3_client.upload_file(outputFile, bucket, 'Processed/' + interviewId + "/" + fileName)
            inputs += " -itsoffset " + str(file["startTimeOffset"]) + " -acodec libopus -i " + inputFile

        mixedFileName = "/mixed-" + interview + ".wav"
        cmd = "ffmpeg -y -loglevel warning" + inputs + " -filter_complex amix=inputs=" + str(len(data["files"])) + ":duration=longest:dropout_transition=3 " + temp_dir + mixedFileName
        p = subprocess.call(cmd, shell=True)  

        #check to see if it's already a multi part audio recording
        key = 'Processed/' + interviewId + "/" + interviewId + "_1.wav"
        objs = list(my_bucket.objects.filter(Prefix=key))
        if len(objs) > 0 and objs[0].key == key:
            count = 3
            #check for interviewid_count.wav to make sure we're not overwriting. When we're not, upload. 
            while(True):
                # check the bucket for the _count file 
                key = 'Processed/' + interviewId + "/" + interviewId + "_" + str(count) + ".wav"
                objs = list(my_bucket.objects.filter(Prefix=key))
                if len(objs) > 0 and objs[0].key == key:
                    #if the key exists, increase the count
                    print(key, " exists")
                else:
                    #if it doesn't, upload the file with that key. 
                    s3_client.upload_file(temp_dir + mixedFileName, bucket, 'Processed/' + interviewId + "/" + interviewId + "_" + str(count) + ".wav")
                    break
                count = count + 1 
        else:
            #if there isn't  a _1, see if there's a .wav
            key = 'Processed/' + interviewId + "/" + interviewId + ".wav"
            objs = list(my_bucket.objects.filter(Prefix=key))
            if len(objs) > 0 and objs[0].key == key:
                #if it does, rename interviewid.wav to interviewid_1.wav
                old_key = 'Processed/' + interviewId + "/" + interviewId + ".wav"
                new_key = key = 'Processed/' + interviewId + "/" + interviewId + "_1.wav"
                s3_resource.Object(bucket,new_key).copy_from(CopySource= bucket + "/" + old_key)
                s3_resource.Object(bucket,old_key).delete()
                s3_client.upload_file(temp_dir + mixedFileName, bucket, 'Processed/' + interviewId + "/" + interviewId + "_2.wav")
            else:
                #upload the file like normal
                s3_client.upload_file(temp_dir + mixedFileName, bucket, 'Processed/' + interviewId + "/" + interviewId + ".wav")

        #delete the non zip files
        for obj in objs:
            if(obj._key[-4:] != ".zip"):
                deletedObj = s3_client.delete_object(Bucket=bucket, Key=obj._key)

    except Exception as e:
        print(e)
        raise e

lambda_handler(zippedKey)